packages/taiko-client-rs/crates/protocol/src/preconfirmation/lookahead/resolver/core.rs

```diff
@@ -0,0 +1,495 @@
+                    let Err(err) = tx.send(LookaheadBroadcast::Blacklisted {
+                        root: event.operatorRegistrationRoot,
+                    })
+                {
+                    warn!(?err, "failed to broadcast blacklist event");
+                }
+            } else if first_topic == Unblacklisted::SIGNATURE_HASH {
+                // Decode and apply unblacklist event.
+                let event =
+                    Unblacklisted::decode_raw_log(log.topics().to_vec(), log.data().data.as_ref())
+                        .map_err(|err| LookaheadError::EventDecode(err.to_string()))?;
+                self.record_unblacklist_event(
+                    event.operatorRegistrationRoot,
+                    event.timestamp.to(),
+                )?;
+                // Broadcast unblacklist update if channel is enabled.
+                if let Some(tx) = &self.broadcast_tx &&
+                    let Err(err) = tx.send(LookaheadBroadcast::Unblacklisted {
+                        root: event.operatorRegistrationRoot,
+                    })
+                {
+                    warn!(?err, "failed to broadcast unblacklist event");
+                }
+            } else {
+                warn!(topic = ?first_topic, "unrecognized log topic");
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Record a blacklist event for an operator and prune history to the allowed lookback window.
+    fn record_blacklist_event(&self, root: B256, at: u64) -> Result<()> {
+        let cutoff = earliest_allowed_timestamp(self.genesis_timestamp)?;
+        let mut timeline = self.blacklist_history.entry(root).or_default();
+        timeline.apply(BlacklistEvent { at, flag: BlacklistFlag::Listed });
+        timeline.prune_before(cutoff);
+        Ok(())
+    }
+
+    /// Record an unblacklist event for an operator and prune history to the allowed lookback
+    /// window.
+    fn record_unblacklist_event(&self, root: B256, at: u64) -> Result<()> {
+        let cutoff = earliest_allowed_timestamp(self.genesis_timestamp)?;
+        let mut timeline = self.blacklist_history.entry(root).or_default();
+        timeline.apply(BlacklistEvent { at, flag: BlacklistFlag::Cleared });
+        timeline.prune_before(cutoff);
+        Ok(())
+    }
+
+    /// Cache a newly observed `LookaheadPosted` event for quick timestamp lookups, capturing the
+    /// whitelist fallback operator at the same block so resolution mirrors on-chain state.
+    pub(crate) async fn store_epoch(
+        &self,
+        event: LookaheadPosted,
+        block_number: Option<u64>,
+    ) -> Result<()> {
+        // Ensure we have the block number for snapshotting.
+        let block = block_number.ok_or_else(|| {
+            LookaheadError::EventDecode("lookahead log missing block number".into())
+        })?;
+
+        // Snapshot whitelist state at the event block state.
+        let fallback_current = self.snapshot_whitelist(block).await?;
```

Comment:
Whitelist fallbacks are snapshotted once per epoch and never refreshed. On-chain LookaheadStore always calls getOperatorForCurrentEpoch() at verification time - is this ok? can mid epoch change happened ?